src/wp-includes/i18n/class-wp-translation-controller.php

```diff
@@ -0,0 +1,404 @@
+
+		if ( false === $translation ) {
+			$text        = $plurals[0];
+			$translation = $this->locate_translation( "{$context}{$text}", $textdomain, $locale );
+
+			if ( false === $translation ) {
+				return false;
+			}
+		}
+
+		/* @var WP_Translation_File $source */
+		$source = $translation['source'];
+		$num    = $source->get_plural_form( $number );
+
+		// See \Translations::translate_plural().
+		return $translation['entries'][ $num ] ?? $translation['entries'][0];
+	}
+
+	/**
+	 * Returns all existing headers for a given text domain.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $textdomain Text domain.
+	 * @return array<string, string> Headers.
+	 */
+	public function get_headers( string $textdomain = 'default' ) {
+		if ( array() === $this->loaded_translations ) {
+			return array();
+		}
+
+		$headers = array();
+
+		foreach ( $this->get_files( $textdomain ) as $moe ) {
+			foreach ( $moe->headers() as $header => $value ) {
+				$headers[ $this->normalize_header( $header ) ] = $value;
+			}
+		}
+
+		return $headers;
+	}
+
+	/**
+	 * Normalizes header names to be capitalized.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $header Header name.
+	 * @return string Normalized header name.
+	 */
+	protected function normalize_header( string $header ) {
+		$parts = explode( '-', $header );
+		$parts = array_map( 'ucfirst', $parts );
+		return implode( '-', $parts );
+	}
+
+	/**
+	 * Returns all entries for a given text domain.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $textdomain Text domain.
+	 * @return array<string, string> Entries.
+	 */
+	public function get_entries( string $textdomain = 'default' ) {
+		if ( array() === $this->loaded_translations ) {
+			return array();
+		}
+
+		$entries = array();
+
+		foreach ( $this->get_files( $textdomain ) as $moe ) {
+			$entries = array_merge( $entries, $moe->entries() );
+		}
+
+		return $entries;
+	}
+
+	/**
+	 * Locates translation for a given string and text domain.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $singular   Singular translation.
+	 * @param string $textdomain Text domain.
+	 * @param string $locale     Optional. Locale. Default current locale.
+	 * @return array{source: WP_Translation_File, entries: string[]}|false Translations on success, false otherwise.
+	 */
+	protected function locate_translation( string $singular, string $textdomain = 'default', string $locale = null ) {
+		if ( array() === $this->loaded_translations ) {
+			return false;
+		}
+
+		// Find the translation in all loaded files for this text domain.
+		foreach ( $this->get_files( $textdomain, $locale ) as $moe ) {
+			$translation = $moe->translate( $singular );
+			if ( false !== $translation ) {
+				return array(
+					'entries' => explode( "\0", $translation ),
+					'source'  => $moe,
+				);
+			}
+			if ( false !== $moe->error() ) {
+				// Unload this file, something is wrong.
+				$this->unload( $textdomain, $moe, $locale );
```

Comment:
Wouldn't we have to call `$l10n_unloaded[ $domain ] = true;` here too for proper compat with that global?

Reply:
While tese share a similar name they're not that related.

Tis controller is self-contained and does not deal with any of the WordPress globals.

Plus this controller holds translations for multiple files and domains, just because one file for a domain is not working anymore does not mean the whole domain needs to be marked as unloaded.

Reply:
@swissspidy I am asking because you set `$l10n_unloaded[ $domain ] = true;` in the other place where you call the `unload()` method? So why is this done in one place but not the other?

Reply:
That other place is within `unload_textdomain()` where this was always there and not a new addition. Here we are in the new translation controller which operates outside of that.

As I said, it is a pretty self-contained library and does not deal with WordPress globals. And even if it would, it also supports multiple translation files _per domain_. And just because loading one file of many failed, it does not mean the whole domain could be marked as unloaded.

Simply put, there is no need to put any `$l10n_unloaded[ $domain ] = true;` calls into this class here as it doesn't make sense.

And if you are wondering what this `$this->unload( $textdomain, $moe, $locale );` line does:

This controller/library is smart and doesn't actually parse translation files until the first translation is looked up. But that also means after loading the translation and looking up the translation, it could detect that the translation file is somehow corrupt, and then remove it from the stack.

That still does not warrant setting that global.